mbed-drivers/v1/I2C.hpp

```diff
@@ -0,0 +1,562 @@
+         * @param[in] irqsafe indicates whether the TransferAdder should use the I2C Object's IRQ-safe allocators
+         */
+        TransferAdder(I2C *i2c, int address, uint32_t hz, bool irqsafe);
+
+        /**
+         * @brief Allocates and constructs a new I2C Segment
+         * @return A new I2C Segment
+         */
+        detail::I2CSegment * new_segment(detail::I2CDirection d);
+
+    public:
+        /**
+         * @brief Set the frequency for this transaction
+         *
+         * By default, the transaction will use the default frequency for the I2C object. This overrides that frequency.
+         * The frequency used will be applied to the whole transaction, not just a single segment.
+         *
+         * @param[in] hz the frequency to set
+         */
+        TransferAdder & frequency(uint32_t hz);
+
+        /**
+         * @brief set an event handler
+         *
+         * An event is triggered when any of the bits in the event mask match the event.
+         * Four event slots are provided, additional event
+         *
+         * @param[in] event the event mask
+         * @param[in] cb the callback to trigger on an event mask match
+         */
+        TransferAdder & on(uint32_t event, const event_callback_t & cb);
+
+        /**
+         * @brief set an event handler
+         *
+         * An event is triggered when any of the bits in the event mask match the event.
+         * Four event slots are provided, additional event
+         *
+         * @param[in] event the event mask
+         * @param[in] cb the callback to trigger on an event mask match
+         */
+        TransferAdder & on(uint32_t event, event_callback_t && cb);
+
+        /**
+         * @brief Queue the transfer
+         *
+         * Hands the transfer over to the resource manager and returns the resource manager's status. No further
+         * configuration of the transfer is possible after apply() has been called.
+         *
+         * @return the error status of submitting the transfer to the resource manager
+         */
+        I2CError apply();
+
+        /**
+         * @brief Add a transmit buffer to the transaction
+         *
+         * @param[in] buf a pointer to the buffer to send
+         * @param[in] len the number of bytes to send
+         */
+        TransferAdder & tx(void *buf, size_t len);
+
+        /**
+         * @brief Add a transmit buffer to the transaction
+         *
+         * @param[in] buf a pointer to and length of the buffer to send
+         */
+        TransferAdder & tx(const Buffer & buf);
+
+        /**
+         * @brief Add an ephermeral transmit buffer to the transaction
+         *
+         * If the buffer is 7 or fewer bytes, it will be managed internally, so the original can be freed.
+         *
+         * @param[in] buf a pointer to the buffer to send
+         * @param[in] len the number of bytes to send
+         */
+        TransferAdder & tx_eph(void *buf, size_t len);
```

Comment:
tx_eph, why not ``tx_epheral()``, Ephemeral class has methods like ``set_ephemeral()``

Same question for void ``set_irq_cb(IRQCallback cb)`` or ``call_irq_cb()`` methods (this one is public API, the former is in detail), we got ``event_callback_t ``.